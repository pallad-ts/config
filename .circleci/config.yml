version: 2.1

orbs:
  node: circleci/node@4.7

jobs:
  main_job:
    executor:
      name: node/default
      tag: "14.18"
    steps:
      - checkout
      - run: "npm ci"
      - run:
          command: ./node_modules/.bin/lerna run compile
          name: compilation
      - run:
          command: ./node_modules/.bin/lerna run lint
          name: linter
      - run:
          command: ./node_modules/.bin/lerna run test
          name: test
      - store_test_results:
          path: ./test_results

workflows:
  main:
    jobs:
      - main_job
