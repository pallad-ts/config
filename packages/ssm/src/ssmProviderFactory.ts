import {SSM} from 'aws-sdk';
import {SSMProvider} from "./SSMProvider";
import DataLoader = require("dataloader");
import {wrapWithDefaultAndTransformer} from '@pallad/config';

const splitArray = require('split-array');

export function parameterDeserializer(parameter: SSM.Parameter) {
    if (parameter.Type === 'StringList') {
        return parameter.Value!.split(',');
    }
    return parameter.Value;
}

function createDataLoader(batchFN: ssmProviderFactory.BatchFunction) {
    return new DataLoader<string, SSMProvider.Value | undefined>(batchFN);
}

export function ssmProviderFactory<TTransformed, TDefault>(
    options?: ssmProviderFactory.Options
) {
    const ssm = options?.ssm ?? new SSM();
    const paramDeserializer = options?.parameterDeserializer ?? parameterDeserializer;

    const batchFn = async (keys: readonly string[]) => {
        const map = new Map<string, SSMProvider.Value | undefined>();
        const batches = splitArray(keys, 10);
        for (const batch of batches) {
            const parameters = await ssm.getParameters({
                Names: batch,
                WithDecryption: true
            }).promise();

            if (parameters.Parameters) {
                for (const parameter of parameters.Parameters) {
                    map.set(parameter.Name!, paramDeserializer(parameter));
                }
            }
        }
        return keys.map(x => map.get(x));
    };

    const dataLoader = (options?.createDataLoader ?? createDataLoader)(batchFn);
    return function <TTransformed, TDefault>(
        key: string,
        opts?: wrapWithDefaultAndTransformer.Options<TTransformed, TDefault, SSMProvider.Value>
    ) {
        const finalKey = (options?.prefix ?? '') + key;
        return wrapWithDefaultAndTransformer(
            new SSMProvider(finalKey, dataLoader),
            opts
        );
    };
}

export namespace ssmProviderFactory {
    export interface Options {
        ssm?: SSM;
        createDataLoader?: (batchFn: BatchFunction) => DataLoader<string, SSMProvider.Value | undefined>;
        parameterDeserializer?: ParameterDeserializer;
        prefix?: string;
    }

    export type BatchFunction = (key: readonly string[]) => Promise<Array<SSMProvider.Value | undefined>>;
    export type ParameterDeserializer = (parameter: SSM.Parameter) => Promise<any> | any;
}
