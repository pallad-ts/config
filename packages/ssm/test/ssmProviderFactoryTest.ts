import * as sinon from 'sinon';
import {type} from '@pallad/config';
import {ssmProviderFactory} from '@src/ssmProviderFactory';

describe('ssmProviderFactory', () => {
    describe('ssm', () => {
        let helper: ReturnType<typeof ssmProviderFactory>;

        const CONFIG = {
            prefix: 'pallad-config-'
        };

        beforeEach(() => {
            helper = ssmProviderFactory({
                ...CONFIG
            });
        });

        describe('loading', () => {
            it('success', () => {
                const dependency = helper('var1');

                return expect(dependency.getValue())
                    .resolves
                    .toEqual('bar')
            });


            it('success with default value', () => {
                const dependency = helper('var100', {default: 'bar'});
                return expect(dependency.getValue())
                    .resolves
                    .toEqual('bar');
            });

            it('failure due to lack of parameter', () => {
                const dependency = helper('var1000');

                return expect(dependency.getValue())
                    .rejects
                    .toMatchSnapshot();
            });

            it('loading with type conversion', () => {
                const dependency = helper('var4', {transformer: type.int});

                return expect(dependency.getValue())
                    .resolves
                    .toEqual(10);
            });

            it('loading array of strings', () => {
                const dependency = helper('var2');
                return expect(dependency.getValue())
                    .resolves
                    .toEqual(['foo', 'bar']);
            });

            it('uses custom dataLoader factory', () => {

                const dataLoader = {};
                const stub = sinon.stub()
                    .resolves(dataLoader);

                ssmProviderFactory({
                    ...CONFIG,
                    createDataLoader: stub,
                });

                sinon.assert.calledOnce(stub);
            });

            it('using parameter deserializer', () => {
                const stub = sinon.stub()
                    .resolves(1000);

                const helper = ssmProviderFactory({
                    ...CONFIG,
                    parameterDeserializer: stub,
                });

                const dependency = helper('var1');

                return expect(dependency.getValue())
                    .resolves
                    .toEqual(1000);
            });

            it('loading many variables', () => {
                const dependencies = [
                    helper('var1'),
                    helper('var2'),
                    helper('var3'),
                    helper('var4'),
                    helper('var5'),
                    helper('var6'),
                    helper('var7'),
                    helper('var8'),
                    helper('var9'),
                    helper('var10'),
                    helper('var11'),
                    helper('var12'),
                ];

                return expect(
                    Promise.all(dependencies.map(x => x.getValue()))
                )
                    .resolves
                    .toEqual([
                        'bar',
                        ['foo', 'bar'],
                        'world',
                        '10',
                        'wjeNdAbmyjtqAU2OcuPR15F4xDzlV1uZHmZNNl1p',
                        "u6yjSXzp9q5yhaCdEWyj",
                        "7FZxN9NHXcgleeP0TSva",
                        "R3aogUWGdvDJy03brMVEZBnqKJYZcp",
                        "XUlAtGjdCwIiHDnojsFqmOC6HO8SyXX",
                        "u97tvnLBUOurP3FcpUWRqbO81GqqH",
                        "HMZNkhmMOxZsylqOaxfNwjeAf9Ju",
                        "DRJeSQn6CmLVuo5eXHGkVABTsgy",
                    ]);
            })
        });
    });
});
