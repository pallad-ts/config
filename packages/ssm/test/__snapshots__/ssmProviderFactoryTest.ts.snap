// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssmProviderFactory ssm loading failure due to lack of parameter 1`] = `[ConfigError: Config value not available: SSM (pallad-config-var1000)]`;
